### ISSUE TITLE
XDR Enums?

### ISSUE THREAD

**Issue by choubacha** [jvff/serde-xdr#2]
Thank you again for adding the data type maps, it really helps! I do have a question though, what data type should I use to express XDR C-like enums?

I thought that I could just make them c-like and it would work:

https://github.com/kbacha/clike-xdr-enums/blob/cd9995d0e8fbd317316e51be132f4ba5749f3f7e/src/main.rs#L5-L17

My thought was that the discriminant value of the enum could be used, but instead it was the ordinal position that was used:

```
$ cargo run
   Compiling test_serde_xdr v0.1.0 (file:///Users/kchoubacha/repos/kbacha/test_serde_xdr)
    Finished dev [unoptimized + debuginfo] target(s) in 0.53 secs
     Running `target/debug/test_serde_xdr`
Ok([0, 0, 0, 0])
Ok([0, 0, 0, 1])
```

One thing that I noticed when playing with enums in rust is that if the enum is all void types, then it's coercible to a i32. Possibly these values could be used as the discriminant then?

Another possibility is a serde derive attribute to set the discriminant value.

Thoughts?

**Comment by jvff** [comment]
Yes, this issue has a deeper problem. Serde doesn't help much here. There's a workaround example using [macros](https://serde.rs/enum-number.html) (except that in this case it should serialize as either `i32` or `u32`).

You can leave this issue open, I'll try to find a more ergonomic solution next week.

Thanks for the request!

**Comment by jvff** [comment]
Wow, I guess the week became a year :sweat_smile: Sorry for taking so long :sweat: 

I believe I have found a workable solution. I created a separate crate ([`serde-enum`]) which can solve this. In theory, all that has to be done is import it and use `#[derive(SerializeEnum, DeserializeEnum)]` instead of the usual `#[derive(Serialize, Deserialize)]`. However, `serde-xdr` also needed a change, so for testing purposes you can add the following to your `Cargo.toml`:

```
[dependencies]
serde = "1.0"
serde-enum = { git = "https://github.com/jvff/serde-enum.git" }
serde-xdr = { git = "https://github.com/jvff/serde-xdr.git", branch = "ignore-enum-variant-names-feature", features = ["ignore-enum-variant-names"] }
```

Let me know if it works. I'll close this issue in a week or so if there are no new developments :)

[`serde-enum`]: https://github.com/jvff/serde-enum
